functions-v2/src/on-analysis-document-imaged.ts

```diff
@@ -0,0 +1,95 @@
+import * as logger from "firebase-functions/logger";
+import * as admin from "firebase-admin";
+import {getAnalysisQueueFirestorePath} from "./utils";
+import {categorizeUrl} from "../lib/src/ai-categorize-document";
+
+// This is one of three functions for AI analysis of documents:
+// 1. Watch for changes to the lastUpdatedAt metadata field and write a queue of docs to process
+// 2. Create screenshots of those documents
+// 3. (This function) Send those screenshots to the AI service for processing, and create comments with the results
+
+// NOTE: these should match the user specified in src/models/stores/user-types.ts
+const commenterName = "Ada Insight";
+const commenterUid = "ada_insight_1";
+
+const imagedQueuePath = getAnalysisQueueFirestorePath("imaged", "{docId}");
+
+async function error(error: string, event: FirestoreEvent<QueryDocumentSnapshot | undefined, Record<string, string>>) {
+  logger.warn("Error processing document", event.document, error);
+  const firestore = admin.firestore();
+  await firestore.doc(getAnalysisQueueFirestorePath("failedAnalyzing", event.params.docId)).set({
+    ...event.data?.data(),
+    error,
+  });
+  await firestore.doc(event.document).delete();
+}
+
+export const onAnalysisDocumentImaged =
+  onDocumentCreated(imagedQueuePath, async (event) => {
+    const {docId} = event.params;
+    const firestore = admin.firestore();
+    const queueDoc = event.data?.data();
+
+    if (queueDoc?.evaluator !== "categorize-design") {
+      await error(`Unexpected value for evaluator: ${queueDoc?.evaluator}`, event);
+      return;
+    }
+
+    const docImageUrl = event.data?.get("docImageUrl");
+
+    const completion = await categorizeUrl(docImageUrl);
+    const reply = completion?.choices[0].message;
+
+    if (reply?.refusal) {
+      await error(`AI refusal: ${reply.refusal}`, event);
+      return;
+    }
+    if (!reply?.parsed) {
+      await error("No response from AI", event);
+      return;
+    }
+    const tags = [reply.parsed.category];
+    const message = reply.parsed.discussion +
+      ` Key Indicators: ${reply.parsed.keyIndicators.join(", ")}`;
+
+    const commentsPath = `demo/AI/documents/${docId}/comments`;
+
+    // Look for existing comment
+    const existing = await firestore.collection(commentsPath)
+      .where("uid", "==", commenterUid).get().then((snapshot) => {
+        if (snapshot.size > 0) {
+          return snapshot.docs[0];
+        } else {
+          return undefined;
+        }
+      });
+
+    if (existing) {
+      logger.info("Updating existing comment for", event.document);
+      await existing.ref.update({
+        tags,
+        content: message,
+        createdAt: admin.firestore.FieldValue.serverTimestamp(),
+      });
+    } else {
+      logger.info("Creating comment for", event.document);
+      // NOTE we are leaving the "network" and "tileId" fields empty in the comment doc.
+      await firestore.collection(commentsPath).add({
```

Comment:
Something general that we are going to have to deal with is that the list of strategies in the Firestore metadata documents are currently updated by the CLUE runtime instead of via a Firebase function that is monitoring the comments. So just adding this comment with a strategy is not going to update the sort work view.

So in a separate story and PR we should add a Firebase function that monitors the comments and updates the Firestore metadata. And in that same PR we can delete the client side code which is updating the strategies. There is a story about handling deleted strategies: https://www.pivotaltracker.com/story/show/188144655 This Firebase function can take care of that too.